# JSON Mini Project
Tamara Monge

Using data in file 'data/world_bank_projects.json' answer the following three questions:

# 1. Find the 10 countries with most projects

In [1]:
# Import pandas, numpy, json, and json_normalize
import pandas as pd
import numpy as np
import json
from pandas.io.json import json_normalize

In [2]:
# Read in the json file as a DataFrame
json_df = pd.read_json('world_bank_projects.json')

First we will simply explore the dataset

In [3]:
# Visually examine the head
json_df.head()

_id  approvalfy board_approval_month  \
0  {'$oid': '52b213b38594d8a2be17c780'}        1999             November   
1  {'$oid': '52b213b38594d8a2be17c781'}        2015             November   
2  {'$oid': '52b213b38594d8a2be17c782'}        2014             November   
3  {'$oid': '52b213b38594d8a2be17c783'}        2014              October   
4  {'$oid': '52b213b38594d8a2be17c784'}        2014              October   

      boardapprovaldate                                 borrower  \
0  2013-11-12T00:00:00Z  FEDERAL DEMOCRATIC REPUBLIC OF ETHIOPIA   
1  2013-11-04T00:00:00Z                    GOVERNMENT OF TUNISIA   
2  2013-11-01T00:00:00Z   MINISTRY OF FINANCE AND ECONOMIC DEVEL   
3  2013-10-31T00:00:00Z   MIN. OF PLANNING AND INT'L COOPERATION   
4  2013-10-31T00:00:00Z                      MINISTRY OF FINANCE   

            closingdate                              country_namecode  \
0  2018-07-07T00:00:00Z  Federal Democratic Republic of Ethiopia!$!ET   
1                   NaN                      Republic of Tunisia!$!TN   
2                   NaN                                   Tuvalu!$!TV   
3                   NaN                        Republic of Yemen!$!RY   
4  2019-04-30T00:00:00Z                       Kingdom of Lesotho!$!LS   

  countrycode                              countryname    countryshortname  \
0          ET  Federal Democratic Republic of Ethiopia            Ethiopia   
1          TN                      Republic of Tunisia             Tunisia   
2          TV                                   Tuvalu              Tuvalu   
3          RY                        Republic of Yemen  Yemen, Republic of   
4          LS                       Kingdom of Lesotho             Lesotho   

                         ...                           sectorcode source  \
0                        ...                          ET,BS,ES,EP   IBRD   
1                        ...                                BZ,BS   IBRD   
2                        ...                                   TI   IBRD   
3                        ...                                   JB   IBRD   
4                        ...                             FH,YW,YZ   IBRD   

   status  supplementprojectflg  \
0  Active                     N   
1  Active                     N   
2  Active                     Y   
3  Active                     N   
4  Active                     N   

                                              theme1  \
0      {'Percent': 100, 'Name': 'Education for all'}   
1  {'Percent': 30, 'Name': 'Other economic manage...   
2    {'Percent': 46, 'Name': 'Regional integration'}   
3  {'Percent': 50, 'Name': 'Participation and civ...   
4  {'Percent': 30, 'Name': 'Export development an...   

                                      theme_namecode    themecode   totalamt  \
0      [{'code': '65', 'name': 'Education for all'}]           65  130000000   
1  [{'code': '24', 'name': 'Other economic manage...        54,24          0   
2  [{'code': '47', 'name': 'Regional integration'...  52,81,25,47    6060000   
3  [{'code': '57', 'name': 'Participation and civ...        59,57          0   
4  [{'code': '45', 'name': 'Export development an...        41,45   13100000   

  totalcommamt                                                url  
0    130000000  http://www.worldbank.org/projects/P129828/ethi...  
1      4700000  http://www.worldbank.org/projects/P144674?lang=en  
2      6060000  http://www.worldbank.org/projects/P145310?lang=en  
3      1500000  http://www.worldbank.org/projects/P144665?lang=en  
4     13100000  http://www.worldbank.org/projects/P144933/seco...  

[5 rows x 50 columns]

In [4]:
# Retrieve dataframe info
json_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 500 entries, 0 to 499
Data columns (total 50 columns):
_id                         500 non-null object
approvalfy                  500 non-null int64
board_approval_month        500 non-null object
boardapprovaldate           500 non-null object
borrower                    485 non-null object
closingdate                 370 non-null object
country_namecode            500 non-null object
countrycode                 500 non-null object
countryname                 500 non-null object
countryshortname            500 non-null object
docty                       446 non-null object
envassesmentcategorycode    430 non-null object
grantamt                    500 non-null int64
ibrdcommamt                 500 non-null int64
id                          500 non-null object
idacommamt                  500 non-null int64
impagency                   472 non-null object
lendinginstr                495 non-null object
lendinginstrtype            495 non

In [5]:
# Retrieve column labels
json_df.columns

Index(['_id', 'approvalfy', 'board_approval_month', 'boardapprovaldate',
       'borrower', 'closingdate', 'country_namecode', 'countrycode',
       'countryname', 'countryshortname', 'docty', 'envassesmentcategorycode',
       'grantamt', 'ibrdcommamt', 'id', 'idacommamt', 'impagency',
       'lendinginstr', 'lendinginstrtype', 'lendprojectcost',
       'majorsector_percent', 'mjsector_namecode', 'mjtheme',
       'mjtheme_namecode', 'mjthemecode', 'prodline', 'prodlinetext',
       'productlinetype', 'project_abstract', 'project_name', 'projectdocs',
       'projectfinancialtype', 'projectstatusdisplay', 'regionname', 'sector',
       'sector1', 'sector2', 'sector3', 'sector4', 'sector_namecode',
       'sectorcode', 'source', 'status', 'supplementprojectflg', 'theme1',
       'theme_namecode', 'themecode', 'totalamt', 'totalcommamt', 'url'],
      dtype='object')

Now, to determine the 10 countries with the most projects, we will use pandas's built-in .value_counts() and .head() methods on the 'countryname' column. 

In [6]:
json_df['countryname'].value_counts().head(10)

People's Republic of China         19
Republic of Indonesia              19
Socialist Republic of Vietnam      17
Republic of India                  16
Republic of Yemen                  13
Nepal                              12
Kingdom of Morocco                 12
People's Republic of Bangladesh    12
Republic of Mozambique             11
Africa                             11
Name: countryname, dtype: int64

# 2. Find the top 10 major project themes (using column 'mjtheme_namecode')

For this question, we will load the json file as a string so that we can use json_normalize to create a dataframe from the nested json string

In [7]:
# Load the json file as a string
json_str = json.load(open('world_bank_projects.json'))

In [8]:
# Use json_normalize to compress the column 'mjtheme_namecode' into a dataframe: theme_df
theme_df = json_normalize(json_str, 'mjtheme_namecode')

In [9]:
# Visually inspect the head
theme_df.head()

code                                   name
0    8                      Human development
1   11                                       
2    1                    Economic management
3    6  Social protection and risk management
4    5                  Trade and integration

Here, we already see there are missing values in the 'name' column.

In [10]:
# Retrieve dataframe info
theme_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1499 entries, 0 to 1498
Data columns (total 2 columns):
code    1499 non-null object
name    1499 non-null object
dtypes: object(2)
memory usage: 23.5+ KB


From the .info() results we can see there are no NaN values, but as we saw in the head, there are missing values. Let's check how many missing values exist in each column.

In [11]:
# Check if there are any empty values in the 'code' column:
theme_df[theme_df['code'] == ''].count()

code    0
name    0
dtype: int64

In [12]:
# Check if there are empty values in the 'name' column:
theme_df[theme_df['name'] == ''].count()

code    122
name    122
dtype: int64

Since the 'name' column has 122 missing values, we will use the 'code' column to determine the top 10 major project themes to avoid skewing the results.

In [13]:
# Pull out the codes corresponding to the 10 most numerous project themes
major_themes_df = pd.DataFrame(theme_df['code'].value_counts().head(10)).reset_index()
major_themes_df.columns = ['code', 'count']
print(major_themes_df)

  code  count
0   11    250
1   10    216
2    8    210
3    2    199
4    6    168
5    4    146
6    7    130
7    5     77
8    9     50
9    1     38


Next, we will determine the theme names associated with each of the 10 major project codes printed above. For this, we will use pd.merge and will need a small dataframe that contains the project theme names and codes and is devoid of missing values and devoid of duplicates.

In [14]:
# Create a new dataframe similar to theme_df except without missing values or duplicates: short_df
short_df = theme_df[theme_df['name'] != ''].drop_duplicates(subset='name')
print(short_df)

    code                                          name
0      8                             Human development
2      1                           Economic management
3      6         Social protection and risk management
4      5                         Trade and integration
5      2                      Public sector governance
6     11  Environment and natural resources management
8      7                   Social dev/gender/inclusion
11     4      Financial and private sector development
18    10                             Rural development
53     9                             Urban development
252    3                                   Rule of law


Finally, we merge the dataframes, thereby adding a column 'name' to major_themes_df

In [15]:
# Add 'name' column to the major_themes_df dataframe by using the pd.merge() function
major_themes_df = pd.merge(major_themes_df, short_df, on='code')
major_themes_df[['count', 'code', 'name']]

count code                                          name
0    250   11  Environment and natural resources management
1    216   10                             Rural development
2    210    8                             Human development
3    199    2                      Public sector governance
4    168    6         Social protection and risk management
5    146    4      Financial and private sector development
6    130    7                   Social dev/gender/inclusion
7     77    5                         Trade and integration
8     50    9                             Urban development
9     38    1                           Economic management

# 3. In 2. above you will notice that some entries have only the code and the name is missing. Create a dataframe with the missing names filled in.

To complete this task, we will sort the dataframe according to code, then according to name so that the missing names are located at the top of each subset and can thus be backfilled later.

In [16]:
# Sort the dataframe by code, then by name and assign to: new_df
new_df = theme_df.sort_values(['code', 'name'])

Now, we replace the missing values from the 'name' column with NaN so that we can backfill them.

In [17]:
# Use a mask to replace the missing values in 'name' column
new_df['name'][new_df['name'] == ''] = np.nan

In [18]:
# Sanity check the number of NaNs
new_df['name'].isnull().sum()

122

Next, we backfill the NaN values

In [19]:
new_df = new_df.fillna(method = 'bfill')

Finally, we confirm that all the NaNs have been filled

In [20]:
# Sanity check the number of NaNs
new_df['name'].isnull().sum()

0

In [21]:
new_df

code                 name
212     1  Economic management
363     1  Economic management
1024    1  Economic management
1114    1  Economic management
1437    1  Economic management
2       1  Economic management
88      1  Economic management
175     1  Economic management
204     1  Economic management
205     1  Economic management
220     1  Economic management
222     1  Economic management
223     1  Economic management
249     1  Economic management
357     1  Economic management
453     1  Economic management
454     1  Economic management
458     1  Economic management
497     1  Economic management
647     1  Economic management
648     1  Economic management
784     1  Economic management
803     1  Economic management
841     1  Economic management
900     1  Economic management
1010    1  Economic management
1045    1  Economic management
1056    1  Economic management
1057    1  Economic management
1078    1  Economic management
...   ...                  ...
597     9    Urban development
598     9    Urban development
650     9    Urban development
669     9    Urban development
732     9    Urban development
737     9    Urban development
738     9    Urban development
739     9    Urban development
742     9    Urban development
822     9    Urban development
917     9    Urban development
930     9    Urban development
940     9    Urban development
962     9    Urban development
1095    9    Urban development
1102    9    Urban development
1103    9    Urban development
1104    9    Urban development
1166    9    Urban development
1290    9    Urban development
1303    9    Urban development
1318    9    Urban development
1342    9    Urban development
1347    9    Urban development
1425    9    Urban development
1426    9    Urban development
1428    9    Urban development
1470    9    Urban development
1473    9    Urban development
1495    9    Urban development

[1499 rows x 2 columns]